File: HW/HW12/Solution.py
class Solution:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        edges = []
        for i in range(n):
            edges.append([0, i + 1, wells[i]])  # Virtual node (0) to house (i+1) with cost wells[i]
        for pipe in pipes:
            edges.append(pipe)
        self.customSort(edges)
        parent = [-1] * (n + 1)
        totalCost = 0
        for edge in edges:
            house1, house2, cost = edge
            parent1 = self.find(parent, house1)
            parent2 = self.find(parent, house2)
            if parent1 != parent2:
                parent[parent1] = parent2
                totalCost += cost
        return totalCost
    def customSort(self, edges):
        edges.sort(key=lambda x: x[2])

    def find(self, parent, x):
        if parent[x] == -1:
            return x
        parent[x] = self.find(parent, parent[x])
        return parent[x]

File: HW/HW12/Solution.java
import java.util.*;

public class Solution {
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        List<int[]> edges = new ArrayList<>();

        for (int i = 0; i < n; i++)
        {
            edges.add(new int[]{0, i + 1, wells[i]});
        }

        for (int[] pipe : pipes)
        {
            edges.add(pipe);
        }

        customSort(edges);

        int[] parent = new int[n + 1];
        Arrays.fill(parent, -1);

        int totalCost = 0;

        for (int[] edge : edges)
        {
            int house1 = edge[0], house2 = edge[1], cost = edge[2];
            int parent1 = find(parent, house1), parent2 = find(parent, house2);

            if (parent1 != parent2)
            {
                parent[parent1] = parent2;
                totalCost += cost;
            }
        }

        return totalCost;
    }

    private void customSort(List<int[]> edges) {
        Collections.sort(edges, new Comparator<int[]>()
        {
            public int compare(int[] a, int[] b)
            {
                return Integer.compare(a[2], b[2]);
            }
        });
    }
    private int find(int[] parent, int x)
    {
        if (parent[x] == -1) return x;
        return parent[x] = find(parent, parent[x]);
    }
}
